File: javascript/day08/part1.js

const fs = require('node:fs');

let input = fs.readFileSync("./input.txt").toString();


const test1 = `RL

AAA = (BBB, CCC)
BBB = (DDD, EEE)
CCC = (ZZZ, GGG)
DDD = (DDD, DDD)
EEE = (EEE, EEE)
GGG = (GGG, GGG)
ZZZ = (ZZZ, ZZZ)`;

const test2 = `LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)`;

const answer1 = 2;
const answer2 = 6;

class Node {
    constructor(main, left, right) {
        this.main = main;
        this.left = left;
        this.rigth = right;
    }
}

function createNodesArray(nodesInput) {
    let nodesArray = [];
    nodesInput.forEach((nodeInput) => {
        const main = nodeInput.slice(0,3);
        const left = nodeInput.slice(7,10);
        const rigth = nodeInput.slice(12,15);
        const nodeItem = new Node(main, left, rigth);
        nodesArray.push(nodeItem);
    });
    return nodesArray;
}


function main(input) {
    const [directions, _, ...nodesInput] = input.split("\n");
    const nodes = createNodesArray(nodesInput);

    let currentNode = nodes.filter((node) => node.main === 'AAA')[0];
    let directionsLength = directions.length;
    let index, count = 0;
    

    while (true) {
        if (currentNode.main === 'ZZZ') {
            break;
        }

        index = count % directionsLength;
        const direction = directions[index];

        if (direction == "R") {
            const nextNode = nodes.filter((node) => node.main === currentNode.rigth)[0];
            currentNode = nextNode;
        } else {
            const nextNode = nodes.filter((node) => node.main === currentNode.left)[0];
            currentNode = nextNode;
        }

        count++;
    }

    return count;
}

const result1 = main(test1);
const result2 = main(test2);
const result = main(input);

console.log(result1);
console.log(result2);
console.log(result);

File: rust/day08/src/part1.rs
#[derive(Copy, Clone)]
struct Node<'a> {
    main: &'a str,
    left: &'a str,
    rigth: &'a str
}

fn create_vec_of_nodes(nodes: Vec<&str>) -> Vec<Node> {
    let mut nodes_vec: Vec<Node> = vec![];
    for node in nodes {
        let main = &node[0..=2];
        let left = &node[7..=9];
        let rigth = &node[12..=14];
        let node_item = Node { main, left, rigth };
        nodes_vec.push(node_item);
    }
    nodes_vec
}

pub fn main(input: &str) -> String  { 
    
    let directions: &str = input.split("\n\n").collect::<Vec<&str>>()[0];
    let nodes: Vec<&str> = input.split("\n\n").collect::<Vec<&str>>()[1].split("\n").collect::<Vec<&str>>();
    let nodes_vec = create_vec_of_nodes(nodes);

    let directions_length = directions.len();
    let mut count: u32 = 0;

    let mut node: &Node = &nodes_vec.into_iter().filter(|x| x.main == "AAA".to_string()).collect::<Vec<Node>>()[0];

    loop {
        let index = (count as usize) % directions_length;
        let direction = directions.chars().nth(index).unwrap();
        node = match direction {
            'R' => nodes_vec.into_iter().filter(|x| x.main == node.rigth).collect::<Vec<Node>>()[0],
            'L' => nodes_vec.into_iter().filter(|x| x.main == node.left).collect::<Vec<Node>>()[0],
            _ => continue,
        };

        if node.main == "ZZZ".to_string() {
            break;
        }

        count += 1;
    }

    count.to_string()


}
